packages/nextcloud_test/lib/src/proxy_http_client.dart

```diff
@@ -0,0 +1,123 @@
+HttpClient getProxyHttpClient({
+  required final void Function(String data) onRequest,
+}) {
+  registerFallbackValue(Uri());
+  registerFallbackValue(ByteStreamFake());
+
+  final mockClient = MockHttpClient();
+  final realClient = HttpClient();
+
+  // ignore: discarded_futures
+  when(() => mockClient.openUrl(any(), any())).thenAnswer((final invocation) async {
+    final mockRequest = MockHttpRequest();
+    final realRequest = await realClient.openUrl(
+      invocation.positionalArguments[0]! as String,
+      invocation.positionalArguments[1]! as Uri,
+    );
+
+    final bodyBytes = BytesBuilder();
+
+    when(() => mockRequest.headers).thenReturn(realRequest.headers);
+    when(() => mockRequest.cookies).thenReturn(realRequest.cookies);
+    when(() => mockRequest.add(any())).thenAnswer((final invocation) {
+      final chunk = invocation.positionalArguments[0] as List<int>;
+      bodyBytes.add(chunk);
+      return realRequest.add(chunk);
+    });
+    when(() => mockRequest.addStream(any())).thenAnswer((final invocation) {
+      final stream = invocation.positionalArguments[0] as Stream<List<int>>;
+      return realRequest.addStream(
+        stream.map((final chunk) {
+          bodyBytes.add(chunk);
+          return chunk;
+        }),
+      );
+    });
+    when(mockRequest.close).thenAnswer((final invocation) async {
+      onRequest(_formatHttpRequest(realRequest, bodyBytes.toBytes()));
+
+      final mockResponse = MockHttpResponse();
+      final realResponse = await realRequest.close();
+
+      when(() => mockResponse.headers).thenReturn(realResponse.headers);
+      when(() => mockResponse.cookies).thenReturn(realResponse.cookies);
+      when(() => mockResponse.contentLength).thenReturn(realResponse.contentLength);
+      when(() => mockResponse.statusCode).thenReturn(realResponse.statusCode);
+      when(() => mockResponse.forEach(any())).thenAnswer(
+        (final invocation) async =>
+            realResponse.forEach(invocation.positionalArguments[0]! as void Function(List<int>)),
+      );
+      when(() => mockResponse.listen(any())).thenAnswer(
+        (final invocation) => realResponse.listen(invocation.positionalArguments[0]! as void Function(List<int>)),
+      );
+      when(() => mockResponse.transform(utf8.decoder)).thenAnswer(
+        (final _) => realResponse.transform(utf8.decoder),
+      );
+      when(() => mockResponse.transform(jsonBytesConverter)).thenAnswer(
+        (final _) => realResponse.transform(jsonBytesConverter),
+      );
+      when(() => mockResponse.transform(xmlBytesConverter)).thenAnswer(
+        (final _) => realResponse.transform(xmlBytesConverter),
+      );
+
+      return mockResponse;
+    });
+
+    return mockRequest;
+  });
+
+  return mockClient;
+}
+
+String _formatHttpRequest(final HttpClientRequest request, final Uint8List body) {
+  final buffer = StringBuffer('${request.method} ${request.uri.replace(port: 80)}');
```

Comment:
in order to work with the other writeln changes this would need to be adapted to also call writeln in a cascade way